src/js/pages/catalog/PackageDetailTab.js

```diff
@@ -78,15 +84,73 @@ class PackageDetailTab extends mixin(StoreMixin) {
     });
   }
 
-  componentDidMount() {
-    super.componentDidMount(...arguments);
+  isSelectedVersionLoading() {
+    const { version } = this.props.location.query;
+    const cosmosPackage = CosmosPackagesStore.getPackageDetails();
+
+    return cosmosPackage.getVersion() !== version;
+  }
 
+  retrievePackageInfo() {
     const { packageName } = this.props.params;
     const { version } = this.props.location.query;
+    const packageVersions = CosmosPackagesStore.getPackageVersions(packageName);
+
+    // fetch package versions available only if not cached
+    if (packageVersions == null) {
+      CosmosPackagesStore.fetchPackageVersions(packageName);
+    }
+
     // Fetch package description
     CosmosPackagesStore.fetchPackageDescription(packageName, version);
   }
 
+  componentDidMount() {
+    super.componentDidMount(...arguments);
+
+    this.retrievePackageInfo();
+  }
+
+  componentWillReceiveProps(nextProps) {
+    const isLoadingVersions =
+      CosmosPackagesStore.getPackageVersions(nextProps.params.packageName) ==
+      null;
+
+    this.setState({ isLoadingVersions });
+  }
+
+  componentDidUpdate(prevProps) {
+    const { isLoadingVersions } = this.state;
+    const { packageName } = this.props.params;
+    const { version } = this.props.location.query;
+
+    // If packageName changed, fetch new data
+    if (prevProps.params.packageName !== packageName) {
+      this.retrievePackageInfo();
+
+      return false;
+    }
+
+    if (version == null && !isLoadingVersions) {
+      // User could navigate here if they don't arrive via the Catalog (which
+      // automatically appends the latest package version to URL).
+      // Redirect to latest version for package
+      const packageVersions = CosmosPackagesStore.getPackageVersions(
+        this.props.params.packageName
+      );
+
+      if (!packageVersions) {
+        return;
+      }
+
+      this.redirectToPackageVersion(packageVersions.getVersions()[0]);
```

Comment:
Why do we need to redirect here?

Reply:
We must have a version in the URL for linking and state management. A user can navigate to `/catalog/packages/kafka` without a version and it will default to selecting the "latest" version of kafka. This redirect ensures we update the URL appropriately else we can be in an inconsistent state.

Reply:
This should be part of the routing logic and shouldn't be handled within a view!